doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -14,223 +14,110 @@
-      .. member:: std::shared_ptr<Block> next
-      .. function:: Block(size_t n)
-      .. function:: char* data()
-
    .. function:: MemArena()
 
       Construct an empty arena.
 
-   .. function:: explicit MemArena(size_t n)
+   .. function:: MemArena(size_t n)
 
-      Construct an arena with :arg:`n` bytes. 
+      Construct an arena with an initial internal block with at least :arg:`n` bytes of free space.
 
    .. function:: MemSpan alloc(size_t n)
 
-      Allocate an :arg:`n` byte chunk of memory in the arena.
+      Allocate memory of size :arg:`n` bytes in the arena.
 
-   .. function:: MemArena& freeze(size_t n = 0)
-
-      Block all further allocation from any existing internal blocks. If :arg:`n` is zero then on the next allocation request a block twice as large as the current generation, otherwise the next internal block will be large enough to hold :arg:`n` bytes.
-
-   .. function:: MemArena& thaw()
+   .. function:: template < typename T, typename ... Args > T * make(Args&& ... args)
 
-      Unallocate all internal blocks that were allocated before the current generation. 
-    
-   .. function:: MemArena& empty()
-     
-      Empties the entire arena and deallocates all underlying memory. Next block size will be equal to the sum of all allocations before the call to empty.
+      Create an instance of :arg:`T`. Memory (:code:`sizeof(T)` bytes) is allocated from the arena and the constructor invoked.
+      This method takes any arguments, which are then passed to the constructor. A pointer to the
+      memory with the newly constructed instance of :arg:`T` is returned. Note if the instance allocates
+      other memory that memory will not be in the arena. Example constructing a :code:`std::string_view` ::
 
-   .. function:: size_t size() const 
+         std::string_view * sv = arena.make<std::string_view>(pointer, n);
 
-      Get the current generation size. The default size of the arena is 32KB unless otherwise specified. 
+   .. function:: MemArena& reserve(size_t n)
 
-   .. function:: size_t remaining() const 
+      Mark the next internal block to have at least :arg:`n` bytes of available space.
 
-      Amount of space left in the generation. 
-
-   .. function:: size_t allocated_size() const
-
-      Total number of bytes allocated in the arena.
-
-   .. function:: size_t unallocated_size() const
-
-      Total number of bytes unallocated in the arena. Can be used to see the internal fragmentation. 
-
-   .. function:: bool contains (void *ptr) const
-
-      Returns whether or not a pointer is in the arena.
-       
-   .. function:: Block* newInternalBlock(size_t n, bool custom)
-
-      Create a new internal block and returns a pointer to the block. 
-
-   .. member:: size_t arena_size
-
-      Current generation size. 
-  
-   .. member:: size_t total_alloc
-
-      Number of bytes allocated out. 
-
-   .. member:: size_t next_block_size
+   .. function:: MemArena& freeze(size_t n = 0)
```

Comment:
It seems like this method could be renamed to `freezeAndReserve`, since it is effectively doing more than just locking down the currently-allocated internal memory blocks.

Reply:
But there is no just `freeze` in contrast, so I think the shorter name is sufficient. There will always be _some_ next internal block size value, it is not the case calling this method create one, it only sets the value.

Reply:
> But there is no just freeze in contrast

But there is a `reserve`!

> There will always be some next internal block size value, it is not the case calling this method create one, it only sets the value.

Yes, this seems clear to me.

```C

size_t newSize = calculateNeededMemForNextLogic();

// Now that I know how much I need, set the internal block size for the next allocation.

arena.reserve(newSize);

// OK, now before we allocate more memory, let's free what we are not going to use.

arena.freeze(); // XXX previous reserve is clobbered here.

// Let's allocate some new memory for the stuff we want to keep around.

arena.alloc(...); // XXX Unexpected internal block size allocated here.

copyMyData(arena, ...);

// We don't need any of the memory that was previously held for the next op.

arena.thaw();
```

The above is a hypothetical example that seems reasonable but where I would not have expected the result.

With the current code, it is better to do

```C
// Prepare to clean up the old but also set the block size for the new.
arena.freeze(newSize);
```

This is what I am saying is confusing. We are passing a size parameter to freeze, which one would think would "freeze `newSize` bytes", but it's really freezing all allocated bytes and then setting some background state only to prepare for a subsequent call by different method.


I am not saying it absolutely must be changed either here or in another PR. I am only observing as a reviewer of changes to documentation that I am seeing for the first time, and this part seemed odd. If it is not that odd to others, then let's go with it.

Reply:
I think it would be better to have a `freeze(bool flag = true)` method for freezing and thawing.  This is what is done with `std::strstream::freeze` https://en.cppreference.com/w/cpp/io/strstream/freeze

`reserve` and `freeze` should be separate methods.